File: src/main/scala/com/vc/hard/L895.scala
package com.vc.hard

class FreqStack() {

  import scala.collection.mutable
  var maxFreq = 0
  val freqMap = new java.util.TreeMap[Int, mutable.Stack[Int]]
  val elementMap = new mutable.HashMap[Int, Int]()

  def push(x: Int) {
    elementMap.put(x, elementMap.getOrElse(x, 0) + 1)
    val freq = elementMap(x)
    maxFreq = Math.max(maxFreq, freq)
    if(!freqMap.containsKey(freq)) {
      freqMap.put(freq, new mutable.Stack[Int]())
    }
    freqMap.get(freq).push(x)
  }

  def pop(): Int = {
    val e = freqMap.get(maxFreq).pop
    elementMap.put(e, elementMap(e) - 1)
    if(freqMap.get(maxFreq).isEmpty) maxFreq -= 1
    e
  }

}

/**
  * Your FreqStack object will be instantiated and called as such:
  * var obj = new FreqStack()
  * obj.push(x)
  * var param_2 = obj.pop()
  */

File: src/main/java/com/vc/hard/L895.java
package com.vc.hard;

import java.util.*;

class FreqStack {

    HashMap<Integer, Integer> count = new HashMap<>();
    HashMap<Integer, Stack<Integer>> value = new HashMap<>();
    int maxCount = 0;

    public FreqStack() {

    }

    public void push(int x) {
        count.put(x, count.getOrDefault(x, 0) + 1);

        int newCount = count.get(x);

        maxCount = Math.max(maxCount, newCount);

        Stack<Integer> st = value.getOrDefault(newCount, new Stack<Integer>());
        st.push(x);

        value.put(newCount, st);
    }

    public int pop() {
        Stack<Integer> st = value.get(maxCount);
        int res = st.pop();
        count.put(res, maxCount - 1);
        if(st.isEmpty()) maxCount--;
        return res;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 */
